// devextreme-angular/ui/nested/custom-shape-dxi.js
"use strict";
/*!
 * devextreme-angular
 * Version: 19.2.3
 * Build date: Tue Oct 22 2019
 *
 * Copyright (c) 2012 - 2019 Developer Express Inc. ALL RIGHTS RESERVED
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file in the root of the project for details.
 *
 * https://github.com/DevExpress/devextreme-angular
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var nested_option_1 = require("../../core/nested-option");
var nested_option_2 = require("../../core/nested-option");
var connection_point_dxi_1 = require("./connection-point-dxi");
var DxiCustomShapeComponent = /** @class */ (function (_super) {
    __extends(DxiCustomShapeComponent, _super);
    function DxiCustomShapeComponent(parentOptionHost, optionHost) {
        var _this = _super.call(this) || this;
        parentOptionHost.setNestedOption(_this);
        optionHost.setHost(_this, _this._fullOptionPath.bind(_this));
        return _this;
    }
    Object.defineProperty(DxiCustomShapeComponent.prototype, "allowEditImage", {
        get: function () {
            return this._getOption('allowEditImage');
        },
        set: function (value) {
            this._setOption('allowEditImage', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxiCustomShapeComponent.prototype, "allowEditText", {
        get: function () {
            return this._getOption('allowEditText');
        },
        set: function (value) {
            this._setOption('allowEditText', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxiCustomShapeComponent.prototype, "backgroundImageHeight", {
        get: function () {
            return this._getOption('backgroundImageHeight');
        },
        set: function (value) {
            this._setOption('backgroundImageHeight', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxiCustomShapeComponent.prototype, "backgroundImageLeft", {
        get: function () {
            return this._getOption('backgroundImageLeft');
        },
        set: function (value) {
            this._setOption('backgroundImageLeft', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxiCustomShapeComponent.prototype, "backgroundImageTop", {
        get: function () {
            return this._getOption('backgroundImageTop');
        },
        set: function (value) {
            this._setOption('backgroundImageTop', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxiCustomShapeComponent.prototype, "backgroundImageUrl", {
        get: function () {
            return this._getOption('backgroundImageUrl');
        },
        set: function (value) {
            this._setOption('backgroundImageUrl', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxiCustomShapeComponent.prototype, "backgroundImageWidth", {
        get: function () {
            return this._getOption('backgroundImageWidth');
        },
        set: function (value) {
            this._setOption('backgroundImageWidth', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxiCustomShapeComponent.prototype, "baseType", {
        get: function () {
            return this._getOption('baseType');
        },
        set: function (value) {
            this._setOption('baseType', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxiCustomShapeComponent.prototype, "category", {
        get: function () {
            return this._getOption('category');
        },
        set: function (value) {
            this._setOption('category', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxiCustomShapeComponent.prototype, "connectionPoints", {
        get: function () {
            return this._getOption('connectionPoints');
        },
        set: function (value) {
            this._setOption('connectionPoints', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxiCustomShapeComponent.prototype, "defaultHeight", {
        get: function () {
            return this._getOption('defaultHeight');
        },
        set: function (value) {
            this._setOption('defaultHeight', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxiCustomShapeComponent.prototype, "defaultImageUrl", {
        get: function () {
            return this._getOption('defaultImageUrl');
        },
        set: function (value) {
            this._setOption('defaultImageUrl', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxiCustomShapeComponent.prototype, "defaultText", {
        get: function () {
            return this._getOption('defaultText');
        },
        set: function (value) {
            this._setOption('defaultText', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxiCustomShapeComponent.prototype, "defaultWidth", {
        get: function () {
            return this._getOption('defaultWidth');
        },
        set: function (value) {
            this._setOption('defaultWidth', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxiCustomShapeComponent.prototype, "imageHeight", {
        get: function () {
            return this._getOption('imageHeight');
        },
        set: function (value) {
            this._setOption('imageHeight', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxiCustomShapeComponent.prototype, "imageLeft", {
        get: function () {
            return this._getOption('imageLeft');
        },
        set: function (value) {
            this._setOption('imageLeft', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxiCustomShapeComponent.prototype, "imageTop", {
        get: function () {
            return this._getOption('imageTop');
        },
        set: function (value) {
            this._setOption('imageTop', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxiCustomShapeComponent.prototype, "imageWidth", {
        get: function () {
            return this._getOption('imageWidth');
        },
        set: function (value) {
            this._setOption('imageWidth', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxiCustomShapeComponent.prototype, "textHeight", {
        get: function () {
            return this._getOption('textHeight');
        },
        set: function (value) {
            this._setOption('textHeight', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxiCustomShapeComponent.prototype, "textLeft", {
        get: function () {
            return this._getOption('textLeft');
        },
        set: function (value) {
            this._setOption('textLeft', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxiCustomShapeComponent.prototype, "textTop", {
        get: function () {
            return this._getOption('textTop');
        },
        set: function (value) {
            this._setOption('textTop', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxiCustomShapeComponent.prototype, "textWidth", {
        get: function () {
            return this._getOption('textWidth');
        },
        set: function (value) {
            this._setOption('textWidth', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxiCustomShapeComponent.prototype, "title", {
        get: function () {
            return this._getOption('title');
        },
        set: function (value) {
            this._setOption('title', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxiCustomShapeComponent.prototype, "type", {
        get: function () {
            return this._getOption('type');
        },
        set: function (value) {
            this._setOption('type', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxiCustomShapeComponent.prototype, "_optionPath", {
        get: function () {
            return 'customShapes';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxiCustomShapeComponent.prototype, "connectionPointsChildren", {
        get: function () {
            return this._getOption('connectionPoints');
        },
        set: function (value) {
            this.setChildren('connectionPoints', value);
        },
        enumerable: true,
        configurable: true
    });
    DxiCustomShapeComponent.decorators = [
        { type: core_1.Component, args: [{
                    selector: 'dxi-custom-shape',
                    template: '',
                    styles: [''],
                    providers: [nested_option_1.NestedOptionHost]
                },] },
    ];
    /** @nocollapse */
    DxiCustomShapeComponent.ctorParameters = function () { return [
        { type: nested_option_1.NestedOptionHost, decorators: [{ type: core_1.SkipSelf }, { type: core_1.Host }] },
        { type: nested_option_1.NestedOptionHost, decorators: [{ type: core_1.Host }] }
    ]; };
    DxiCustomShapeComponent.propDecorators = {
        allowEditImage: [{ type: core_1.Input }],
        allowEditText: [{ type: core_1.Input }],
        backgroundImageHeight: [{ type: core_1.Input }],
        backgroundImageLeft: [{ type: core_1.Input }],
        backgroundImageTop: [{ type: core_1.Input }],
        backgroundImageUrl: [{ type: core_1.Input }],
        backgroundImageWidth: [{ type: core_1.Input }],
        baseType: [{ type: core_1.Input }],
        category: [{ type: core_1.Input }],
        connectionPoints: [{ type: core_1.Input }],
        defaultHeight: [{ type: core_1.Input }],
        defaultImageUrl: [{ type: core_1.Input }],
        defaultText: [{ type: core_1.Input }],
        defaultWidth: [{ type: core_1.Input }],
        imageHeight: [{ type: core_1.Input }],
        imageLeft: [{ type: core_1.Input }],
        imageTop: [{ type: core_1.Input }],
        imageWidth: [{ type: core_1.Input }],
        textHeight: [{ type: core_1.Input }],
        textLeft: [{ type: core_1.Input }],
        textTop: [{ type: core_1.Input }],
        textWidth: [{ type: core_1.Input }],
        title: [{ type: core_1.Input }],
        type: [{ type: core_1.Input }],
        connectionPointsChildren: [{ type: core_1.ContentChildren, args: [core_1.forwardRef(function () { return connection_point_dxi_1.DxiConnectionPointComponent; }),] }]
    };
    return DxiCustomShapeComponent;
}(nested_option_2.CollectionNestedOption));
exports.DxiCustomShapeComponent = DxiCustomShapeComponent;
var DxiCustomShapeModule = /** @class */ (function () {
    function DxiCustomShapeModule() {
    }
    DxiCustomShapeModule.decorators = [
        { type: core_1.NgModule, args: [{
                    declarations: [
                        DxiCustomShapeComponent
                    ],
                    exports: [
                        DxiCustomShapeComponent
                    ],
                },] },
    ];
    return DxiCustomShapeModule;
}());
exports.DxiCustomShapeModule = DxiCustomShapeModule;
