// devextreme-angular/ui/nested/item-dragging.js
"use strict";
/*!
 * devextreme-angular
 * Version: 19.2.3
 * Build date: Tue Oct 22 2019
 *
 * Copyright (c) 2012 - 2019 Developer Express Inc. ALL RIGHTS RESERVED
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file in the root of the project for details.
 *
 * https://github.com/DevExpress/devextreme-angular
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var nested_option_1 = require("../../core/nested-option");
var sortable_options_1 = require("./base/sortable-options");
var DxoItemDraggingComponent = /** @class */ (function (_super) {
    __extends(DxoItemDraggingComponent, _super);
    function DxoItemDraggingComponent(parentOptionHost, optionHost) {
        var _this = _super.call(this) || this;
        parentOptionHost.setNestedOption(_this);
        optionHost.setHost(_this, _this._fullOptionPath.bind(_this));
        return _this;
    }
    Object.defineProperty(DxoItemDraggingComponent.prototype, "_optionPath", {
        get: function () {
            return 'itemDragging';
        },
        enumerable: true,
        configurable: true
    });
    DxoItemDraggingComponent.decorators = [
        { type: core_1.Component, args: [{
                    selector: 'dxo-item-dragging',
                    template: '',
                    styles: [''],
                    providers: [nested_option_1.NestedOptionHost],
                    inputs: [
                        'allowDropInsideItem',
                        'allowReordering',
                        'autoScroll',
                        'boundary',
                        'container',
                        'cursorOffset',
                        'data',
                        'dragDirection',
                        'dragTemplate',
                        'dropFeedbackMode',
                        'elementAttr',
                        'filter',
                        'group',
                        'handle',
                        'height',
                        'itemOrientation',
                        'moveItemOnDrop',
                        'onAdd',
                        'onDisposing',
                        'onDragChange',
                        'onDragEnd',
                        'onDragMove',
                        'onDragStart',
                        'onInitialized',
                        'onOptionChanged',
                        'onRemove',
                        'onReorder',
                        'rtlEnabled',
                        'scrollSensitivity',
                        'scrollSpeed',
                        'width'
                    ]
                },] },
    ];
    /** @nocollapse */
    DxoItemDraggingComponent.ctorParameters = function () { return [
        { type: nested_option_1.NestedOptionHost, decorators: [{ type: core_1.SkipSelf }, { type: core_1.Host }] },
        { type: nested_option_1.NestedOptionHost, decorators: [{ type: core_1.Host }] }
    ]; };
    return DxoItemDraggingComponent;
}(sortable_options_1.DxoSortableOptions));
exports.DxoItemDraggingComponent = DxoItemDraggingComponent;
var DxoItemDraggingModule = /** @class */ (function () {
    function DxoItemDraggingModule() {
    }
    DxoItemDraggingModule.decorators = [
        { type: core_1.NgModule, args: [{
                    declarations: [
                        DxoItemDraggingComponent
                    ],
                    exports: [
                        DxoItemDraggingComponent
                    ],
                },] },
    ];
    return DxoItemDraggingModule;
}());
exports.DxoItemDraggingModule = DxoItemDraggingModule;
