// devextreme-angular/ui/nested/nodes.js
"use strict";
/*!
 * devextreme-angular
 * Version: 19.2.3
 * Build date: Tue Oct 22 2019
 *
 * Copyright (c) 2012 - 2019 Developer Express Inc. ALL RIGHTS RESERVED
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file in the root of the project for details.
 *
 * https://github.com/DevExpress/devextreme-angular
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var nested_option_1 = require("../../core/nested-option");
var nested_option_2 = require("../../core/nested-option");
var DxoNodesComponent = /** @class */ (function (_super) {
    __extends(DxoNodesComponent, _super);
    function DxoNodesComponent(parentOptionHost, optionHost) {
        var _this = _super.call(this) || this;
        parentOptionHost.setNestedOption(_this);
        optionHost.setHost(_this, _this._fullOptionPath.bind(_this));
        return _this;
    }
    Object.defineProperty(DxoNodesComponent.prototype, "autoLayout", {
        get: function () {
            return this._getOption('autoLayout');
        },
        set: function (value) {
            this._setOption('autoLayout', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoNodesComponent.prototype, "childrenExpr", {
        get: function () {
            return this._getOption('childrenExpr');
        },
        set: function (value) {
            this._setOption('childrenExpr', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoNodesComponent.prototype, "containerKeyExpr", {
        get: function () {
            return this._getOption('containerKeyExpr');
        },
        set: function (value) {
            this._setOption('containerKeyExpr', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoNodesComponent.prototype, "dataSource", {
        get: function () {
            return this._getOption('dataSource');
        },
        set: function (value) {
            this._setOption('dataSource', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoNodesComponent.prototype, "heightExpr", {
        get: function () {
            return this._getOption('heightExpr');
        },
        set: function (value) {
            this._setOption('heightExpr', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoNodesComponent.prototype, "imageUrlExpr", {
        get: function () {
            return this._getOption('imageUrlExpr');
        },
        set: function (value) {
            this._setOption('imageUrlExpr', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoNodesComponent.prototype, "itemsExpr", {
        get: function () {
            return this._getOption('itemsExpr');
        },
        set: function (value) {
            this._setOption('itemsExpr', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoNodesComponent.prototype, "keyExpr", {
        get: function () {
            return this._getOption('keyExpr');
        },
        set: function (value) {
            this._setOption('keyExpr', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoNodesComponent.prototype, "leftExpr", {
        get: function () {
            return this._getOption('leftExpr');
        },
        set: function (value) {
            this._setOption('leftExpr', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoNodesComponent.prototype, "lockedExpr", {
        get: function () {
            return this._getOption('lockedExpr');
        },
        set: function (value) {
            this._setOption('lockedExpr', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoNodesComponent.prototype, "parentKeyExpr", {
        get: function () {
            return this._getOption('parentKeyExpr');
        },
        set: function (value) {
            this._setOption('parentKeyExpr', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoNodesComponent.prototype, "styleExpr", {
        get: function () {
            return this._getOption('styleExpr');
        },
        set: function (value) {
            this._setOption('styleExpr', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoNodesComponent.prototype, "textExpr", {
        get: function () {
            return this._getOption('textExpr');
        },
        set: function (value) {
            this._setOption('textExpr', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoNodesComponent.prototype, "textStyleExpr", {
        get: function () {
            return this._getOption('textStyleExpr');
        },
        set: function (value) {
            this._setOption('textStyleExpr', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoNodesComponent.prototype, "topExpr", {
        get: function () {
            return this._getOption('topExpr');
        },
        set: function (value) {
            this._setOption('topExpr', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoNodesComponent.prototype, "typeExpr", {
        get: function () {
            return this._getOption('typeExpr');
        },
        set: function (value) {
            this._setOption('typeExpr', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoNodesComponent.prototype, "widthExpr", {
        get: function () {
            return this._getOption('widthExpr');
        },
        set: function (value) {
            this._setOption('widthExpr', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoNodesComponent.prototype, "zIndexExpr", {
        get: function () {
            return this._getOption('zIndexExpr');
        },
        set: function (value) {
            this._setOption('zIndexExpr', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoNodesComponent.prototype, "_optionPath", {
        get: function () {
            return 'nodes';
        },
        enumerable: true,
        configurable: true
    });
    DxoNodesComponent.decorators = [
        { type: core_1.Component, args: [{
                    selector: 'dxo-nodes',
                    template: '',
                    styles: [''],
                    providers: [nested_option_1.NestedOptionHost]
                },] },
    ];
    /** @nocollapse */
    DxoNodesComponent.ctorParameters = function () { return [
        { type: nested_option_1.NestedOptionHost, decorators: [{ type: core_1.SkipSelf }, { type: core_1.Host }] },
        { type: nested_option_1.NestedOptionHost, decorators: [{ type: core_1.Host }] }
    ]; };
    DxoNodesComponent.propDecorators = {
        autoLayout: [{ type: core_1.Input }],
        childrenExpr: [{ type: core_1.Input }],
        containerKeyExpr: [{ type: core_1.Input }],
        dataSource: [{ type: core_1.Input }],
        heightExpr: [{ type: core_1.Input }],
        imageUrlExpr: [{ type: core_1.Input }],
        itemsExpr: [{ type: core_1.Input }],
        keyExpr: [{ type: core_1.Input }],
        leftExpr: [{ type: core_1.Input }],
        lockedExpr: [{ type: core_1.Input }],
        parentKeyExpr: [{ type: core_1.Input }],
        styleExpr: [{ type: core_1.Input }],
        textExpr: [{ type: core_1.Input }],
        textStyleExpr: [{ type: core_1.Input }],
        topExpr: [{ type: core_1.Input }],
        typeExpr: [{ type: core_1.Input }],
        widthExpr: [{ type: core_1.Input }],
        zIndexExpr: [{ type: core_1.Input }]
    };
    return DxoNodesComponent;
}(nested_option_2.NestedOption));
exports.DxoNodesComponent = DxoNodesComponent;
var DxoNodesModule = /** @class */ (function () {
    function DxoNodesModule() {
    }
    DxoNodesModule.decorators = [
        { type: core_1.NgModule, args: [{
                    declarations: [
                        DxoNodesComponent
                    ],
                    exports: [
                        DxoNodesComponent
                    ],
                },] },
    ];
    return DxoNodesModule;
}());
exports.DxoNodesModule = DxoNodesModule;
