// devextreme-angular/ui/nested/page-size.js
"use strict";
/*!
 * devextreme-angular
 * Version: 19.2.3
 * Build date: Tue Oct 22 2019
 *
 * Copyright (c) 2012 - 2019 Developer Express Inc. ALL RIGHTS RESERVED
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file in the root of the project for details.
 *
 * https://github.com/DevExpress/devextreme-angular
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var nested_option_1 = require("../../core/nested-option");
var nested_option_2 = require("../../core/nested-option");
var item_dxi_1 = require("./item-dxi");
var DxoPageSizeComponent = /** @class */ (function (_super) {
    __extends(DxoPageSizeComponent, _super);
    function DxoPageSizeComponent(parentOptionHost, optionHost) {
        var _this = _super.call(this) || this;
        parentOptionHost.setNestedOption(_this);
        optionHost.setHost(_this, _this._fullOptionPath.bind(_this));
        return _this;
    }
    Object.defineProperty(DxoPageSizeComponent.prototype, "height", {
        get: function () {
            return this._getOption('height');
        },
        set: function (value) {
            this._setOption('height', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoPageSizeComponent.prototype, "items", {
        get: function () {
            return this._getOption('items');
        },
        set: function (value) {
            this._setOption('items', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoPageSizeComponent.prototype, "width", {
        get: function () {
            return this._getOption('width');
        },
        set: function (value) {
            this._setOption('width', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoPageSizeComponent.prototype, "_optionPath", {
        get: function () {
            return 'pageSize';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoPageSizeComponent.prototype, "itemsChildren", {
        get: function () {
            return this._getOption('items');
        },
        set: function (value) {
            this.setChildren('items', value);
        },
        enumerable: true,
        configurable: true
    });
    DxoPageSizeComponent.decorators = [
        { type: core_1.Component, args: [{
                    selector: 'dxo-page-size',
                    template: '',
                    styles: [''],
                    providers: [nested_option_1.NestedOptionHost]
                },] },
    ];
    /** @nocollapse */
    DxoPageSizeComponent.ctorParameters = function () { return [
        { type: nested_option_1.NestedOptionHost, decorators: [{ type: core_1.SkipSelf }, { type: core_1.Host }] },
        { type: nested_option_1.NestedOptionHost, decorators: [{ type: core_1.Host }] }
    ]; };
    DxoPageSizeComponent.propDecorators = {
        height: [{ type: core_1.Input }],
        items: [{ type: core_1.Input }],
        width: [{ type: core_1.Input }],
        itemsChildren: [{ type: core_1.ContentChildren, args: [core_1.forwardRef(function () { return item_dxi_1.DxiItemComponent; }),] }]
    };
    return DxoPageSizeComponent;
}(nested_option_2.NestedOption));
exports.DxoPageSizeComponent = DxoPageSizeComponent;
var DxoPageSizeModule = /** @class */ (function () {
    function DxoPageSizeModule() {
    }
    DxoPageSizeModule.decorators = [
        { type: core_1.NgModule, args: [{
                    declarations: [
                        DxoPageSizeComponent
                    ],
                    exports: [
                        DxoPageSizeComponent
                    ],
                },] },
    ];
    return DxoPageSizeModule;
}());
exports.DxoPageSizeModule = DxoPageSizeModule;
