// devextreme-angular/ui/nested/permissions.js
"use strict";
/*!
 * devextreme-angular
 * Version: 19.2.3
 * Build date: Tue Oct 22 2019
 *
 * Copyright (c) 2012 - 2019 Developer Express Inc. ALL RIGHTS RESERVED
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file in the root of the project for details.
 *
 * https://github.com/DevExpress/devextreme-angular
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var nested_option_1 = require("../../core/nested-option");
var nested_option_2 = require("../../core/nested-option");
var DxoPermissionsComponent = /** @class */ (function (_super) {
    __extends(DxoPermissionsComponent, _super);
    function DxoPermissionsComponent(parentOptionHost, optionHost) {
        var _this = _super.call(this) || this;
        parentOptionHost.setNestedOption(_this);
        optionHost.setHost(_this, _this._fullOptionPath.bind(_this));
        return _this;
    }
    Object.defineProperty(DxoPermissionsComponent.prototype, "copy", {
        get: function () {
            return this._getOption('copy');
        },
        set: function (value) {
            this._setOption('copy', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoPermissionsComponent.prototype, "create", {
        get: function () {
            return this._getOption('create');
        },
        set: function (value) {
            this._setOption('create', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoPermissionsComponent.prototype, "download", {
        get: function () {
            return this._getOption('download');
        },
        set: function (value) {
            this._setOption('download', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoPermissionsComponent.prototype, "move", {
        get: function () {
            return this._getOption('move');
        },
        set: function (value) {
            this._setOption('move', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoPermissionsComponent.prototype, "remove", {
        get: function () {
            return this._getOption('remove');
        },
        set: function (value) {
            this._setOption('remove', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoPermissionsComponent.prototype, "rename", {
        get: function () {
            return this._getOption('rename');
        },
        set: function (value) {
            this._setOption('rename', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoPermissionsComponent.prototype, "upload", {
        get: function () {
            return this._getOption('upload');
        },
        set: function (value) {
            this._setOption('upload', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoPermissionsComponent.prototype, "_optionPath", {
        get: function () {
            return 'permissions';
        },
        enumerable: true,
        configurable: true
    });
    DxoPermissionsComponent.decorators = [
        { type: core_1.Component, args: [{
                    selector: 'dxo-permissions',
                    template: '',
                    styles: [''],
                    providers: [nested_option_1.NestedOptionHost]
                },] },
    ];
    /** @nocollapse */
    DxoPermissionsComponent.ctorParameters = function () { return [
        { type: nested_option_1.NestedOptionHost, decorators: [{ type: core_1.SkipSelf }, { type: core_1.Host }] },
        { type: nested_option_1.NestedOptionHost, decorators: [{ type: core_1.Host }] }
    ]; };
    DxoPermissionsComponent.propDecorators = {
        copy: [{ type: core_1.Input }],
        create: [{ type: core_1.Input }],
        download: [{ type: core_1.Input }],
        move: [{ type: core_1.Input }],
        remove: [{ type: core_1.Input }],
        rename: [{ type: core_1.Input }],
        upload: [{ type: core_1.Input }]
    };
    return DxoPermissionsComponent;
}(nested_option_2.NestedOption));
exports.DxoPermissionsComponent = DxoPermissionsComponent;
var DxoPermissionsModule = /** @class */ (function () {
    function DxoPermissionsModule() {
    }
    DxoPermissionsModule.decorators = [
        { type: core_1.NgModule, args: [{
                    declarations: [
                        DxoPermissionsComponent
                    ],
                    exports: [
                        DxoPermissionsComponent
                    ],
                },] },
    ];
    return DxoPermissionsModule;
}());
exports.DxoPermissionsModule = DxoPermissionsModule;
