// devextreme-angular/ui/nested/properties-panel.js
"use strict";
/*!
 * devextreme-angular
 * Version: 19.2.3
 * Build date: Tue Oct 22 2019
 *
 * Copyright (c) 2012 - 2019 Developer Express Inc. ALL RIGHTS RESERVED
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file in the root of the project for details.
 *
 * https://github.com/DevExpress/devextreme-angular
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var nested_option_1 = require("../../core/nested-option");
var nested_option_2 = require("../../core/nested-option");
var group_dxi_1 = require("./group-dxi");
var DxoPropertiesPanelComponent = /** @class */ (function (_super) {
    __extends(DxoPropertiesPanelComponent, _super);
    function DxoPropertiesPanelComponent(parentOptionHost, optionHost) {
        var _this = _super.call(this) || this;
        parentOptionHost.setNestedOption(_this);
        optionHost.setHost(_this, _this._fullOptionPath.bind(_this));
        return _this;
    }
    Object.defineProperty(DxoPropertiesPanelComponent.prototype, "collapsible", {
        get: function () {
            return this._getOption('collapsible');
        },
        set: function (value) {
            this._setOption('collapsible', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoPropertiesPanelComponent.prototype, "enabled", {
        get: function () {
            return this._getOption('enabled');
        },
        set: function (value) {
            this._setOption('enabled', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoPropertiesPanelComponent.prototype, "groups", {
        get: function () {
            return this._getOption('groups');
        },
        set: function (value) {
            this._setOption('groups', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoPropertiesPanelComponent.prototype, "_optionPath", {
        get: function () {
            return 'propertiesPanel';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoPropertiesPanelComponent.prototype, "groupsChildren", {
        get: function () {
            return this._getOption('groups');
        },
        set: function (value) {
            this.setChildren('groups', value);
        },
        enumerable: true,
        configurable: true
    });
    DxoPropertiesPanelComponent.decorators = [
        { type: core_1.Component, args: [{
                    selector: 'dxo-properties-panel',
                    template: '',
                    styles: [''],
                    providers: [nested_option_1.NestedOptionHost]
                },] },
    ];
    /** @nocollapse */
    DxoPropertiesPanelComponent.ctorParameters = function () { return [
        { type: nested_option_1.NestedOptionHost, decorators: [{ type: core_1.SkipSelf }, { type: core_1.Host }] },
        { type: nested_option_1.NestedOptionHost, decorators: [{ type: core_1.Host }] }
    ]; };
    DxoPropertiesPanelComponent.propDecorators = {
        collapsible: [{ type: core_1.Input }],
        enabled: [{ type: core_1.Input }],
        groups: [{ type: core_1.Input }],
        groupsChildren: [{ type: core_1.ContentChildren, args: [core_1.forwardRef(function () { return group_dxi_1.DxiGroupComponent; }),] }]
    };
    return DxoPropertiesPanelComponent;
}(nested_option_2.NestedOption));
exports.DxoPropertiesPanelComponent = DxoPropertiesPanelComponent;
var DxoPropertiesPanelModule = /** @class */ (function () {
    function DxoPropertiesPanelModule() {
    }
    DxoPropertiesPanelModule.decorators = [
        { type: core_1.NgModule, args: [{
                    declarations: [
                        DxoPropertiesPanelComponent
                    ],
                    exports: [
                        DxoPropertiesPanelComponent
                    ],
                },] },
    ];
    return DxoPropertiesPanelModule;
}());
exports.DxoPropertiesPanelModule = DxoPropertiesPanelModule;
