// devextreme-angular/ui/nested/zoom-level.js
"use strict";
/*!
 * devextreme-angular
 * Version: 19.2.3
 * Build date: Tue Oct 22 2019
 *
 * Copyright (c) 2012 - 2019 Developer Express Inc. ALL RIGHTS RESERVED
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file in the root of the project for details.
 *
 * https://github.com/DevExpress/devextreme-angular
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var nested_option_1 = require("../../core/nested-option");
var nested_option_2 = require("../../core/nested-option");
var DxoZoomLevelComponent = /** @class */ (function (_super) {
    __extends(DxoZoomLevelComponent, _super);
    function DxoZoomLevelComponent(parentOptionHost, optionHost) {
        var _this = _super.call(this) || this;
        parentOptionHost.setNestedOption(_this);
        optionHost.setHost(_this, _this._fullOptionPath.bind(_this));
        return _this;
    }
    Object.defineProperty(DxoZoomLevelComponent.prototype, "items", {
        get: function () {
            return this._getOption('items');
        },
        set: function (value) {
            this._setOption('items', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoZoomLevelComponent.prototype, "value", {
        get: function () {
            return this._getOption('value');
        },
        set: function (value) {
            this._setOption('value', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoZoomLevelComponent.prototype, "_optionPath", {
        get: function () {
            return 'zoomLevel';
        },
        enumerable: true,
        configurable: true
    });
    DxoZoomLevelComponent.decorators = [
        { type: core_1.Component, args: [{
                    selector: 'dxo-zoom-level',
                    template: '',
                    styles: [''],
                    providers: [nested_option_1.NestedOptionHost]
                },] },
    ];
    /** @nocollapse */
    DxoZoomLevelComponent.ctorParameters = function () { return [
        { type: nested_option_1.NestedOptionHost, decorators: [{ type: core_1.SkipSelf }, { type: core_1.Host }] },
        { type: nested_option_1.NestedOptionHost, decorators: [{ type: core_1.Host }] }
    ]; };
    DxoZoomLevelComponent.propDecorators = {
        items: [{ type: core_1.Input }],
        value: [{ type: core_1.Input }]
    };
    return DxoZoomLevelComponent;
}(nested_option_2.NestedOption));
exports.DxoZoomLevelComponent = DxoZoomLevelComponent;
var DxoZoomLevelModule = /** @class */ (function () {
    function DxoZoomLevelModule() {
    }
    DxoZoomLevelModule.decorators = [
        { type: core_1.NgModule, args: [{
                    declarations: [
                        DxoZoomLevelComponent
                    ],
                    exports: [
                        DxoZoomLevelComponent
                    ],
                },] },
    ];
    return DxoZoomLevelModule;
}());
exports.DxoZoomLevelModule = DxoZoomLevelModule;
