// devextreme-angular/esm5/ui/nested/appointment-dragging.js
/*!
 * devextreme-angular
 * Version: 19.2.3
 * Build date: Tue Oct 22 2019
 *
 * Copyright (c) 2012 - 2019 Developer Express Inc. ALL RIGHTS RESERVED
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file in the root of the project for details.
 *
 * https://github.com/DevExpress/devextreme-angular
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
/* tslint:disable:max-line-length */
import { Component, NgModule, Host, SkipSelf, Input } from '@angular/core';
import { NestedOptionHost } from '../../core/nested-option';
import { NestedOption } from '../../core/nested-option';
var DxoAppointmentDraggingComponent = /** @class */ (function (_super) {
    __extends(DxoAppointmentDraggingComponent, _super);
    function DxoAppointmentDraggingComponent(parentOptionHost, optionHost) {
        var _this = _super.call(this) || this;
        parentOptionHost.setNestedOption(_this);
        optionHost.setHost(_this, _this._fullOptionPath.bind(_this));
        return _this;
    }
    Object.defineProperty(DxoAppointmentDraggingComponent.prototype, "autoScroll", {
        get: function () {
            return this._getOption('autoScroll');
        },
        set: function (value) {
            this._setOption('autoScroll', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoAppointmentDraggingComponent.prototype, "data", {
        get: function () {
            return this._getOption('data');
        },
        set: function (value) {
            this._setOption('data', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoAppointmentDraggingComponent.prototype, "group", {
        get: function () {
            return this._getOption('group');
        },
        set: function (value) {
            this._setOption('group', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoAppointmentDraggingComponent.prototype, "onAdd", {
        get: function () {
            return this._getOption('onAdd');
        },
        set: function (value) {
            this._setOption('onAdd', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoAppointmentDraggingComponent.prototype, "onDragEnd", {
        get: function () {
            return this._getOption('onDragEnd');
        },
        set: function (value) {
            this._setOption('onDragEnd', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoAppointmentDraggingComponent.prototype, "onDragMove", {
        get: function () {
            return this._getOption('onDragMove');
        },
        set: function (value) {
            this._setOption('onDragMove', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoAppointmentDraggingComponent.prototype, "onDragStart", {
        get: function () {
            return this._getOption('onDragStart');
        },
        set: function (value) {
            this._setOption('onDragStart', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoAppointmentDraggingComponent.prototype, "onRemove", {
        get: function () {
            return this._getOption('onRemove');
        },
        set: function (value) {
            this._setOption('onRemove', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoAppointmentDraggingComponent.prototype, "scrollSensitivity", {
        get: function () {
            return this._getOption('scrollSensitivity');
        },
        set: function (value) {
            this._setOption('scrollSensitivity', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoAppointmentDraggingComponent.prototype, "scrollSpeed", {
        get: function () {
            return this._getOption('scrollSpeed');
        },
        set: function (value) {
            this._setOption('scrollSpeed', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoAppointmentDraggingComponent.prototype, "_optionPath", {
        get: function () {
            return 'appointmentDragging';
        },
        enumerable: true,
        configurable: true
    });
    DxoAppointmentDraggingComponent.decorators = [
        { type: Component, args: [{
                    selector: 'dxo-appointment-dragging',
                    template: '',
                    styles: [''],
                    providers: [NestedOptionHost]
                },] },
    ];
    /** @nocollapse */
    DxoAppointmentDraggingComponent.ctorParameters = function () { return [
        { type: NestedOptionHost, decorators: [{ type: SkipSelf }, { type: Host }] },
        { type: NestedOptionHost, decorators: [{ type: Host }] }
    ]; };
    DxoAppointmentDraggingComponent.propDecorators = {
        autoScroll: [{ type: Input }],
        data: [{ type: Input }],
        group: [{ type: Input }],
        onAdd: [{ type: Input }],
        onDragEnd: [{ type: Input }],
        onDragMove: [{ type: Input }],
        onDragStart: [{ type: Input }],
        onRemove: [{ type: Input }],
        scrollSensitivity: [{ type: Input }],
        scrollSpeed: [{ type: Input }]
    };
    return DxoAppointmentDraggingComponent;
}(NestedOption));
export { DxoAppointmentDraggingComponent };
var DxoAppointmentDraggingModule = /** @class */ (function () {
    function DxoAppointmentDraggingModule() {
    }
    DxoAppointmentDraggingModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        DxoAppointmentDraggingComponent
                    ],
                    exports: [
                        DxoAppointmentDraggingComponent
                    ],
                },] },
    ];
    return DxoAppointmentDraggingModule;
}());
export { DxoAppointmentDraggingModule };
