// devextreme-angular/esm5/ui/nested/auto-layout.js
/*!
 * devextreme-angular
 * Version: 19.2.3
 * Build date: Tue Oct 22 2019
 *
 * Copyright (c) 2012 - 2019 Developer Express Inc. ALL RIGHTS RESERVED
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file in the root of the project for details.
 *
 * https://github.com/DevExpress/devextreme-angular
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
/* tslint:disable:max-line-length */
import { Component, NgModule, Host, SkipSelf, Input } from '@angular/core';
import { NestedOptionHost } from '../../core/nested-option';
import { NestedOption } from '../../core/nested-option';
var DxoAutoLayoutComponent = /** @class */ (function (_super) {
    __extends(DxoAutoLayoutComponent, _super);
    function DxoAutoLayoutComponent(parentOptionHost, optionHost) {
        var _this = _super.call(this) || this;
        parentOptionHost.setNestedOption(_this);
        optionHost.setHost(_this, _this._fullOptionPath.bind(_this));
        return _this;
    }
    Object.defineProperty(DxoAutoLayoutComponent.prototype, "orientation", {
        get: function () {
            return this._getOption('orientation');
        },
        set: function (value) {
            this._setOption('orientation', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoAutoLayoutComponent.prototype, "type", {
        get: function () {
            return this._getOption('type');
        },
        set: function (value) {
            this._setOption('type', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoAutoLayoutComponent.prototype, "_optionPath", {
        get: function () {
            return 'autoLayout';
        },
        enumerable: true,
        configurable: true
    });
    DxoAutoLayoutComponent.decorators = [
        { type: Component, args: [{
                    selector: 'dxo-auto-layout',
                    template: '',
                    styles: [''],
                    providers: [NestedOptionHost]
                },] },
    ];
    /** @nocollapse */
    DxoAutoLayoutComponent.ctorParameters = function () { return [
        { type: NestedOptionHost, decorators: [{ type: SkipSelf }, { type: Host }] },
        { type: NestedOptionHost, decorators: [{ type: Host }] }
    ]; };
    DxoAutoLayoutComponent.propDecorators = {
        orientation: [{ type: Input }],
        type: [{ type: Input }]
    };
    return DxoAutoLayoutComponent;
}(NestedOption));
export { DxoAutoLayoutComponent };
var DxoAutoLayoutModule = /** @class */ (function () {
    function DxoAutoLayoutModule() {
    }
    DxoAutoLayoutModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        DxoAutoLayoutComponent
                    ],
                    exports: [
                        DxoAutoLayoutComponent
                    ],
                },] },
    ];
    return DxoAutoLayoutModule;
}());
export { DxoAutoLayoutModule };
