// devextreme-angular/esm5/ui/nested/connection-point-dxi.js
/*!
 * devextreme-angular
 * Version: 19.2.3
 * Build date: Tue Oct 22 2019
 *
 * Copyright (c) 2012 - 2019 Developer Express Inc. ALL RIGHTS RESERVED
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file in the root of the project for details.
 *
 * https://github.com/DevExpress/devextreme-angular
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
/* tslint:disable:max-line-length */
import { Component, NgModule, Host, SkipSelf, Input } from '@angular/core';
import { NestedOptionHost } from '../../core/nested-option';
import { CollectionNestedOption } from '../../core/nested-option';
var DxiConnectionPointComponent = /** @class */ (function (_super) {
    __extends(DxiConnectionPointComponent, _super);
    function DxiConnectionPointComponent(parentOptionHost, optionHost) {
        var _this = _super.call(this) || this;
        parentOptionHost.setNestedOption(_this);
        optionHost.setHost(_this, _this._fullOptionPath.bind(_this));
        return _this;
    }
    Object.defineProperty(DxiConnectionPointComponent.prototype, "x", {
        get: function () {
            return this._getOption('x');
        },
        set: function (value) {
            this._setOption('x', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxiConnectionPointComponent.prototype, "y", {
        get: function () {
            return this._getOption('y');
        },
        set: function (value) {
            this._setOption('y', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxiConnectionPointComponent.prototype, "_optionPath", {
        get: function () {
            return 'connectionPoints';
        },
        enumerable: true,
        configurable: true
    });
    DxiConnectionPointComponent.decorators = [
        { type: Component, args: [{
                    selector: 'dxi-connection-point',
                    template: '',
                    styles: [''],
                    providers: [NestedOptionHost]
                },] },
    ];
    /** @nocollapse */
    DxiConnectionPointComponent.ctorParameters = function () { return [
        { type: NestedOptionHost, decorators: [{ type: SkipSelf }, { type: Host }] },
        { type: NestedOptionHost, decorators: [{ type: Host }] }
    ]; };
    DxiConnectionPointComponent.propDecorators = {
        x: [{ type: Input }],
        y: [{ type: Input }]
    };
    return DxiConnectionPointComponent;
}(CollectionNestedOption));
export { DxiConnectionPointComponent };
var DxiConnectionPointModule = /** @class */ (function () {
    function DxiConnectionPointModule() {
    }
    DxiConnectionPointModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        DxiConnectionPointComponent
                    ],
                    exports: [
                        DxiConnectionPointComponent
                    ],
                },] },
    ];
    return DxiConnectionPointModule;
}());
export { DxiConnectionPointModule };
