// devextreme-angular/esm5/ui/nested/dependencies.js
/*!
 * devextreme-angular
 * Version: 19.2.3
 * Build date: Tue Oct 22 2019
 *
 * Copyright (c) 2012 - 2019 Developer Express Inc. ALL RIGHTS RESERVED
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file in the root of the project for details.
 *
 * https://github.com/DevExpress/devextreme-angular
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
/* tslint:disable:max-line-length */
import { Component, NgModule, Host, SkipSelf, Input } from '@angular/core';
import { NestedOptionHost } from '../../core/nested-option';
import { NestedOption } from '../../core/nested-option';
var DxoDependenciesComponent = /** @class */ (function (_super) {
    __extends(DxoDependenciesComponent, _super);
    function DxoDependenciesComponent(parentOptionHost, optionHost) {
        var _this = _super.call(this) || this;
        parentOptionHost.setNestedOption(_this);
        optionHost.setHost(_this, _this._fullOptionPath.bind(_this));
        return _this;
    }
    Object.defineProperty(DxoDependenciesComponent.prototype, "dataSource", {
        get: function () {
            return this._getOption('dataSource');
        },
        set: function (value) {
            this._setOption('dataSource', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoDependenciesComponent.prototype, "keyExpr", {
        get: function () {
            return this._getOption('keyExpr');
        },
        set: function (value) {
            this._setOption('keyExpr', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoDependenciesComponent.prototype, "predecessorIdExpr", {
        get: function () {
            return this._getOption('predecessorIdExpr');
        },
        set: function (value) {
            this._setOption('predecessorIdExpr', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoDependenciesComponent.prototype, "successorIdExpr", {
        get: function () {
            return this._getOption('successorIdExpr');
        },
        set: function (value) {
            this._setOption('successorIdExpr', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoDependenciesComponent.prototype, "typeExpr", {
        get: function () {
            return this._getOption('typeExpr');
        },
        set: function (value) {
            this._setOption('typeExpr', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoDependenciesComponent.prototype, "_optionPath", {
        get: function () {
            return 'dependencies';
        },
        enumerable: true,
        configurable: true
    });
    DxoDependenciesComponent.decorators = [
        { type: Component, args: [{
                    selector: 'dxo-dependencies',
                    template: '',
                    styles: [''],
                    providers: [NestedOptionHost]
                },] },
    ];
    /** @nocollapse */
    DxoDependenciesComponent.ctorParameters = function () { return [
        { type: NestedOptionHost, decorators: [{ type: SkipSelf }, { type: Host }] },
        { type: NestedOptionHost, decorators: [{ type: Host }] }
    ]; };
    DxoDependenciesComponent.propDecorators = {
        dataSource: [{ type: Input }],
        keyExpr: [{ type: Input }],
        predecessorIdExpr: [{ type: Input }],
        successorIdExpr: [{ type: Input }],
        typeExpr: [{ type: Input }]
    };
    return DxoDependenciesComponent;
}(NestedOption));
export { DxoDependenciesComponent };
var DxoDependenciesModule = /** @class */ (function () {
    function DxoDependenciesModule() {
    }
    DxoDependenciesModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        DxoDependenciesComponent
                    ],
                    exports: [
                        DxoDependenciesComponent
                    ],
                },] },
    ];
    return DxoDependenciesModule;
}());
export { DxoDependenciesModule };
