// devextreme-angular/esm5/ui/nested/edges.js
/*!
 * devextreme-angular
 * Version: 19.2.3
 * Build date: Tue Oct 22 2019
 *
 * Copyright (c) 2012 - 2019 Developer Express Inc. ALL RIGHTS RESERVED
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file in the root of the project for details.
 *
 * https://github.com/DevExpress/devextreme-angular
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
/* tslint:disable:max-line-length */
import { Component, NgModule, Host, SkipSelf, Input } from '@angular/core';
import { NestedOptionHost } from '../../core/nested-option';
import { NestedOption } from '../../core/nested-option';
var DxoEdgesComponent = /** @class */ (function (_super) {
    __extends(DxoEdgesComponent, _super);
    function DxoEdgesComponent(parentOptionHost, optionHost) {
        var _this = _super.call(this) || this;
        parentOptionHost.setNestedOption(_this);
        optionHost.setHost(_this, _this._fullOptionPath.bind(_this));
        return _this;
    }
    Object.defineProperty(DxoEdgesComponent.prototype, "dataSource", {
        get: function () {
            return this._getOption('dataSource');
        },
        set: function (value) {
            this._setOption('dataSource', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoEdgesComponent.prototype, "fromExpr", {
        get: function () {
            return this._getOption('fromExpr');
        },
        set: function (value) {
            this._setOption('fromExpr', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoEdgesComponent.prototype, "fromLineEndExpr", {
        get: function () {
            return this._getOption('fromLineEndExpr');
        },
        set: function (value) {
            this._setOption('fromLineEndExpr', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoEdgesComponent.prototype, "fromPointIndexExpr", {
        get: function () {
            return this._getOption('fromPointIndexExpr');
        },
        set: function (value) {
            this._setOption('fromPointIndexExpr', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoEdgesComponent.prototype, "keyExpr", {
        get: function () {
            return this._getOption('keyExpr');
        },
        set: function (value) {
            this._setOption('keyExpr', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoEdgesComponent.prototype, "lineTypeExpr", {
        get: function () {
            return this._getOption('lineTypeExpr');
        },
        set: function (value) {
            this._setOption('lineTypeExpr', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoEdgesComponent.prototype, "lockedExpr", {
        get: function () {
            return this._getOption('lockedExpr');
        },
        set: function (value) {
            this._setOption('lockedExpr', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoEdgesComponent.prototype, "pointsExpr", {
        get: function () {
            return this._getOption('pointsExpr');
        },
        set: function (value) {
            this._setOption('pointsExpr', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoEdgesComponent.prototype, "styleExpr", {
        get: function () {
            return this._getOption('styleExpr');
        },
        set: function (value) {
            this._setOption('styleExpr', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoEdgesComponent.prototype, "textExpr", {
        get: function () {
            return this._getOption('textExpr');
        },
        set: function (value) {
            this._setOption('textExpr', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoEdgesComponent.prototype, "textStyleExpr", {
        get: function () {
            return this._getOption('textStyleExpr');
        },
        set: function (value) {
            this._setOption('textStyleExpr', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoEdgesComponent.prototype, "toExpr", {
        get: function () {
            return this._getOption('toExpr');
        },
        set: function (value) {
            this._setOption('toExpr', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoEdgesComponent.prototype, "toLineEndExpr", {
        get: function () {
            return this._getOption('toLineEndExpr');
        },
        set: function (value) {
            this._setOption('toLineEndExpr', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoEdgesComponent.prototype, "toPointIndexExpr", {
        get: function () {
            return this._getOption('toPointIndexExpr');
        },
        set: function (value) {
            this._setOption('toPointIndexExpr', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoEdgesComponent.prototype, "zIndexExpr", {
        get: function () {
            return this._getOption('zIndexExpr');
        },
        set: function (value) {
            this._setOption('zIndexExpr', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoEdgesComponent.prototype, "_optionPath", {
        get: function () {
            return 'edges';
        },
        enumerable: true,
        configurable: true
    });
    DxoEdgesComponent.decorators = [
        { type: Component, args: [{
                    selector: 'dxo-edges',
                    template: '',
                    styles: [''],
                    providers: [NestedOptionHost]
                },] },
    ];
    /** @nocollapse */
    DxoEdgesComponent.ctorParameters = function () { return [
        { type: NestedOptionHost, decorators: [{ type: SkipSelf }, { type: Host }] },
        { type: NestedOptionHost, decorators: [{ type: Host }] }
    ]; };
    DxoEdgesComponent.propDecorators = {
        dataSource: [{ type: Input }],
        fromExpr: [{ type: Input }],
        fromLineEndExpr: [{ type: Input }],
        fromPointIndexExpr: [{ type: Input }],
        keyExpr: [{ type: Input }],
        lineTypeExpr: [{ type: Input }],
        lockedExpr: [{ type: Input }],
        pointsExpr: [{ type: Input }],
        styleExpr: [{ type: Input }],
        textExpr: [{ type: Input }],
        textStyleExpr: [{ type: Input }],
        toExpr: [{ type: Input }],
        toLineEndExpr: [{ type: Input }],
        toPointIndexExpr: [{ type: Input }],
        zIndexExpr: [{ type: Input }]
    };
    return DxoEdgesComponent;
}(NestedOption));
export { DxoEdgesComponent };
var DxoEdgesModule = /** @class */ (function () {
    function DxoEdgesModule() {
    }
    DxoEdgesModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        DxoEdgesComponent
                    ],
                    exports: [
                        DxoEdgesComponent
                    ],
                },] },
    ];
    return DxoEdgesModule;
}());
export { DxoEdgesModule };
