// devextreme-angular/esm5/ui/nested/group-dxi.js
/*!
 * devextreme-angular
 * Version: 19.2.3
 * Build date: Tue Oct 22 2019
 *
 * Copyright (c) 2012 - 2019 Developer Express Inc. ALL RIGHTS RESERVED
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file in the root of the project for details.
 *
 * https://github.com/DevExpress/devextreme-angular
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
/* tslint:disable:max-line-length */
import { Component, NgModule, Host, SkipSelf, Input } from '@angular/core';
import { NestedOptionHost } from '../../core/nested-option';
import { CollectionNestedOption } from '../../core/nested-option';
var DxiGroupComponent = /** @class */ (function (_super) {
    __extends(DxiGroupComponent, _super);
    function DxiGroupComponent(parentOptionHost, optionHost) {
        var _this = _super.call(this) || this;
        parentOptionHost.setNestedOption(_this);
        optionHost.setHost(_this, _this._fullOptionPath.bind(_this));
        return _this;
    }
    Object.defineProperty(DxiGroupComponent.prototype, "commands", {
        get: function () {
            return this._getOption('commands');
        },
        set: function (value) {
            this._setOption('commands', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxiGroupComponent.prototype, "category", {
        get: function () {
            return this._getOption('category');
        },
        set: function (value) {
            this._setOption('category', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxiGroupComponent.prototype, "displayMode", {
        get: function () {
            return this._getOption('displayMode');
        },
        set: function (value) {
            this._setOption('displayMode', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxiGroupComponent.prototype, "expanded", {
        get: function () {
            return this._getOption('expanded');
        },
        set: function (value) {
            this._setOption('expanded', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxiGroupComponent.prototype, "shapes", {
        get: function () {
            return this._getOption('shapes');
        },
        set: function (value) {
            this._setOption('shapes', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxiGroupComponent.prototype, "title", {
        get: function () {
            return this._getOption('title');
        },
        set: function (value) {
            this._setOption('title', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxiGroupComponent.prototype, "_optionPath", {
        get: function () {
            return 'groups';
        },
        enumerable: true,
        configurable: true
    });
    DxiGroupComponent.decorators = [
        { type: Component, args: [{
                    selector: 'dxi-group',
                    template: '',
                    styles: [''],
                    providers: [NestedOptionHost]
                },] },
    ];
    /** @nocollapse */
    DxiGroupComponent.ctorParameters = function () { return [
        { type: NestedOptionHost, decorators: [{ type: SkipSelf }, { type: Host }] },
        { type: NestedOptionHost, decorators: [{ type: Host }] }
    ]; };
    DxiGroupComponent.propDecorators = {
        commands: [{ type: Input }],
        category: [{ type: Input }],
        displayMode: [{ type: Input }],
        expanded: [{ type: Input }],
        shapes: [{ type: Input }],
        title: [{ type: Input }]
    };
    return DxiGroupComponent;
}(CollectionNestedOption));
export { DxiGroupComponent };
var DxiGroupModule = /** @class */ (function () {
    function DxiGroupModule() {
    }
    DxiGroupModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        DxiGroupComponent
                    ],
                    exports: [
                        DxiGroupComponent
                    ],
                },] },
    ];
    return DxiGroupModule;
}());
export { DxiGroupModule };
