// devextreme-angular/esm5/ui/nested/item-view.js
/*!
 * devextreme-angular
 * Version: 19.2.3
 * Build date: Tue Oct 22 2019
 *
 * Copyright (c) 2012 - 2019 Developer Express Inc. ALL RIGHTS RESERVED
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file in the root of the project for details.
 *
 * https://github.com/DevExpress/devextreme-angular
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
/* tslint:disable:max-line-length */
import { Component, NgModule, Host, SkipSelf, Input } from '@angular/core';
import { NestedOptionHost } from '../../core/nested-option';
import { NestedOption } from '../../core/nested-option';
var DxoItemViewComponent = /** @class */ (function (_super) {
    __extends(DxoItemViewComponent, _super);
    function DxoItemViewComponent(parentOptionHost, optionHost) {
        var _this = _super.call(this) || this;
        parentOptionHost.setNestedOption(_this);
        optionHost.setHost(_this, _this._fullOptionPath.bind(_this));
        return _this;
    }
    Object.defineProperty(DxoItemViewComponent.prototype, "mode", {
        get: function () {
            return this._getOption('mode');
        },
        set: function (value) {
            this._setOption('mode', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoItemViewComponent.prototype, "showFolders", {
        get: function () {
            return this._getOption('showFolders');
        },
        set: function (value) {
            this._setOption('showFolders', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoItemViewComponent.prototype, "showParentFolder", {
        get: function () {
            return this._getOption('showParentFolder');
        },
        set: function (value) {
            this._setOption('showParentFolder', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoItemViewComponent.prototype, "_optionPath", {
        get: function () {
            return 'itemView';
        },
        enumerable: true,
        configurable: true
    });
    DxoItemViewComponent.decorators = [
        { type: Component, args: [{
                    selector: 'dxo-item-view',
                    template: '',
                    styles: [''],
                    providers: [NestedOptionHost]
                },] },
    ];
    /** @nocollapse */
    DxoItemViewComponent.ctorParameters = function () { return [
        { type: NestedOptionHost, decorators: [{ type: SkipSelf }, { type: Host }] },
        { type: NestedOptionHost, decorators: [{ type: Host }] }
    ]; };
    DxoItemViewComponent.propDecorators = {
        mode: [{ type: Input }],
        showFolders: [{ type: Input }],
        showParentFolder: [{ type: Input }]
    };
    return DxoItemViewComponent;
}(NestedOption));
export { DxoItemViewComponent };
var DxoItemViewModule = /** @class */ (function () {
    function DxoItemViewModule() {
    }
    DxoItemViewModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        DxoItemViewComponent
                    ],
                    exports: [
                        DxoItemViewComponent
                    ],
                },] },
    ];
    return DxoItemViewModule;
}());
export { DxoItemViewModule };
