// devextreme-angular/esm5/ui/nested/mention-dxi.js
/*!
 * devextreme-angular
 * Version: 19.2.3
 * Build date: Tue Oct 22 2019
 *
 * Copyright (c) 2012 - 2019 Developer Express Inc. ALL RIGHTS RESERVED
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file in the root of the project for details.
 *
 * https://github.com/DevExpress/devextreme-angular
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
/* tslint:disable:max-line-length */
/* tslint:disable:use-input-property-decorator */
import { Component, NgModule, Host, SkipSelf } from '@angular/core';
import { NestedOptionHost } from '../../core/nested-option';
import { DxiHtmlEditorMention } from './base/html-editor-mention-dxi';
var DxiMentionComponent = /** @class */ (function (_super) {
    __extends(DxiMentionComponent, _super);
    function DxiMentionComponent(parentOptionHost, optionHost) {
        var _this = _super.call(this) || this;
        parentOptionHost.setNestedOption(_this);
        optionHost.setHost(_this, _this._fullOptionPath.bind(_this));
        return _this;
    }
    Object.defineProperty(DxiMentionComponent.prototype, "_optionPath", {
        get: function () {
            return 'mentions';
        },
        enumerable: true,
        configurable: true
    });
    DxiMentionComponent.decorators = [
        { type: Component, args: [{
                    selector: 'dxi-mention',
                    template: '',
                    styles: [''],
                    providers: [NestedOptionHost],
                    inputs: [
                        'dataSource',
                        'displayExpr',
                        'itemTemplate',
                        'marker',
                        'minSearchLength',
                        'searchExpr',
                        'searchTimeout',
                        'template',
                        'valueExpr'
                    ]
                },] },
    ];
    /** @nocollapse */
    DxiMentionComponent.ctorParameters = function () { return [
        { type: NestedOptionHost, decorators: [{ type: SkipSelf }, { type: Host }] },
        { type: NestedOptionHost, decorators: [{ type: Host }] }
    ]; };
    return DxiMentionComponent;
}(DxiHtmlEditorMention));
export { DxiMentionComponent };
var DxiMentionModule = /** @class */ (function () {
    function DxiMentionModule() {
    }
    DxiMentionModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        DxiMentionComponent
                    ],
                    exports: [
                        DxiMentionComponent
                    ],
                },] },
    ];
    return DxiMentionModule;
}());
export { DxiMentionModule };
