// devextreme-angular/esm5/ui/nested/resource-assignments.js
/*!
 * devextreme-angular
 * Version: 19.2.3
 * Build date: Tue Oct 22 2019
 *
 * Copyright (c) 2012 - 2019 Developer Express Inc. ALL RIGHTS RESERVED
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file in the root of the project for details.
 *
 * https://github.com/DevExpress/devextreme-angular
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
/* tslint:disable:max-line-length */
import { Component, NgModule, Host, SkipSelf, Input } from '@angular/core';
import { NestedOptionHost } from '../../core/nested-option';
import { NestedOption } from '../../core/nested-option';
var DxoResourceAssignmentsComponent = /** @class */ (function (_super) {
    __extends(DxoResourceAssignmentsComponent, _super);
    function DxoResourceAssignmentsComponent(parentOptionHost, optionHost) {
        var _this = _super.call(this) || this;
        parentOptionHost.setNestedOption(_this);
        optionHost.setHost(_this, _this._fullOptionPath.bind(_this));
        return _this;
    }
    Object.defineProperty(DxoResourceAssignmentsComponent.prototype, "dataSource", {
        get: function () {
            return this._getOption('dataSource');
        },
        set: function (value) {
            this._setOption('dataSource', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoResourceAssignmentsComponent.prototype, "keyExpr", {
        get: function () {
            return this._getOption('keyExpr');
        },
        set: function (value) {
            this._setOption('keyExpr', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoResourceAssignmentsComponent.prototype, "resourceIdExpr", {
        get: function () {
            return this._getOption('resourceIdExpr');
        },
        set: function (value) {
            this._setOption('resourceIdExpr', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoResourceAssignmentsComponent.prototype, "taskIdExpr", {
        get: function () {
            return this._getOption('taskIdExpr');
        },
        set: function (value) {
            this._setOption('taskIdExpr', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoResourceAssignmentsComponent.prototype, "_optionPath", {
        get: function () {
            return 'resourceAssignments';
        },
        enumerable: true,
        configurable: true
    });
    DxoResourceAssignmentsComponent.decorators = [
        { type: Component, args: [{
                    selector: 'dxo-resource-assignments',
                    template: '',
                    styles: [''],
                    providers: [NestedOptionHost]
                },] },
    ];
    /** @nocollapse */
    DxoResourceAssignmentsComponent.ctorParameters = function () { return [
        { type: NestedOptionHost, decorators: [{ type: SkipSelf }, { type: Host }] },
        { type: NestedOptionHost, decorators: [{ type: Host }] }
    ]; };
    DxoResourceAssignmentsComponent.propDecorators = {
        dataSource: [{ type: Input }],
        keyExpr: [{ type: Input }],
        resourceIdExpr: [{ type: Input }],
        taskIdExpr: [{ type: Input }]
    };
    return DxoResourceAssignmentsComponent;
}(NestedOption));
export { DxoResourceAssignmentsComponent };
var DxoResourceAssignmentsModule = /** @class */ (function () {
    function DxoResourceAssignmentsModule() {
    }
    DxoResourceAssignmentsModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        DxoResourceAssignmentsComponent
                    ],
                    exports: [
                        DxoResourceAssignmentsComponent
                    ],
                },] },
    ];
    return DxoResourceAssignmentsModule;
}());
export { DxoResourceAssignmentsModule };
