// devextreme-angular/esm5/ui/nested/row-dragging.js
/*!
 * devextreme-angular
 * Version: 19.2.3
 * Build date: Tue Oct 22 2019
 *
 * Copyright (c) 2012 - 2019 Developer Express Inc. ALL RIGHTS RESERVED
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file in the root of the project for details.
 *
 * https://github.com/DevExpress/devextreme-angular
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
/* tslint:disable:max-line-length */
import { Component, NgModule, Host, SkipSelf, Input } from '@angular/core';
import { NestedOptionHost } from '../../core/nested-option';
import { NestedOption } from '../../core/nested-option';
var DxoRowDraggingComponent = /** @class */ (function (_super) {
    __extends(DxoRowDraggingComponent, _super);
    function DxoRowDraggingComponent(parentOptionHost, optionHost) {
        var _this = _super.call(this) || this;
        parentOptionHost.setNestedOption(_this);
        optionHost.setHost(_this, _this._fullOptionPath.bind(_this));
        return _this;
    }
    Object.defineProperty(DxoRowDraggingComponent.prototype, "allowDropInsideItem", {
        get: function () {
            return this._getOption('allowDropInsideItem');
        },
        set: function (value) {
            this._setOption('allowDropInsideItem', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoRowDraggingComponent.prototype, "allowReordering", {
        get: function () {
            return this._getOption('allowReordering');
        },
        set: function (value) {
            this._setOption('allowReordering', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoRowDraggingComponent.prototype, "autoScroll", {
        get: function () {
            return this._getOption('autoScroll');
        },
        set: function (value) {
            this._setOption('autoScroll', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoRowDraggingComponent.prototype, "boundary", {
        get: function () {
            return this._getOption('boundary');
        },
        set: function (value) {
            this._setOption('boundary', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoRowDraggingComponent.prototype, "container", {
        get: function () {
            return this._getOption('container');
        },
        set: function (value) {
            this._setOption('container', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoRowDraggingComponent.prototype, "cursorOffset", {
        get: function () {
            return this._getOption('cursorOffset');
        },
        set: function (value) {
            this._setOption('cursorOffset', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoRowDraggingComponent.prototype, "data", {
        get: function () {
            return this._getOption('data');
        },
        set: function (value) {
            this._setOption('data', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoRowDraggingComponent.prototype, "dragDirection", {
        get: function () {
            return this._getOption('dragDirection');
        },
        set: function (value) {
            this._setOption('dragDirection', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoRowDraggingComponent.prototype, "dragTemplate", {
        get: function () {
            return this._getOption('dragTemplate');
        },
        set: function (value) {
            this._setOption('dragTemplate', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoRowDraggingComponent.prototype, "dropFeedbackMode", {
        get: function () {
            return this._getOption('dropFeedbackMode');
        },
        set: function (value) {
            this._setOption('dropFeedbackMode', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoRowDraggingComponent.prototype, "filter", {
        get: function () {
            return this._getOption('filter');
        },
        set: function (value) {
            this._setOption('filter', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoRowDraggingComponent.prototype, "group", {
        get: function () {
            return this._getOption('group');
        },
        set: function (value) {
            this._setOption('group', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoRowDraggingComponent.prototype, "handle", {
        get: function () {
            return this._getOption('handle');
        },
        set: function (value) {
            this._setOption('handle', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoRowDraggingComponent.prototype, "onAdd", {
        get: function () {
            return this._getOption('onAdd');
        },
        set: function (value) {
            this._setOption('onAdd', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoRowDraggingComponent.prototype, "onDragChange", {
        get: function () {
            return this._getOption('onDragChange');
        },
        set: function (value) {
            this._setOption('onDragChange', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoRowDraggingComponent.prototype, "onDragEnd", {
        get: function () {
            return this._getOption('onDragEnd');
        },
        set: function (value) {
            this._setOption('onDragEnd', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoRowDraggingComponent.prototype, "onDragMove", {
        get: function () {
            return this._getOption('onDragMove');
        },
        set: function (value) {
            this._setOption('onDragMove', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoRowDraggingComponent.prototype, "onDragStart", {
        get: function () {
            return this._getOption('onDragStart');
        },
        set: function (value) {
            this._setOption('onDragStart', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoRowDraggingComponent.prototype, "onRemove", {
        get: function () {
            return this._getOption('onRemove');
        },
        set: function (value) {
            this._setOption('onRemove', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoRowDraggingComponent.prototype, "onReorder", {
        get: function () {
            return this._getOption('onReorder');
        },
        set: function (value) {
            this._setOption('onReorder', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoRowDraggingComponent.prototype, "scrollSensitivity", {
        get: function () {
            return this._getOption('scrollSensitivity');
        },
        set: function (value) {
            this._setOption('scrollSensitivity', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoRowDraggingComponent.prototype, "scrollSpeed", {
        get: function () {
            return this._getOption('scrollSpeed');
        },
        set: function (value) {
            this._setOption('scrollSpeed', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoRowDraggingComponent.prototype, "showDragIcons", {
        get: function () {
            return this._getOption('showDragIcons');
        },
        set: function (value) {
            this._setOption('showDragIcons', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoRowDraggingComponent.prototype, "_optionPath", {
        get: function () {
            return 'rowDragging';
        },
        enumerable: true,
        configurable: true
    });
    DxoRowDraggingComponent.decorators = [
        { type: Component, args: [{
                    selector: 'dxo-row-dragging',
                    template: '',
                    styles: [''],
                    providers: [NestedOptionHost]
                },] },
    ];
    /** @nocollapse */
    DxoRowDraggingComponent.ctorParameters = function () { return [
        { type: NestedOptionHost, decorators: [{ type: SkipSelf }, { type: Host }] },
        { type: NestedOptionHost, decorators: [{ type: Host }] }
    ]; };
    DxoRowDraggingComponent.propDecorators = {
        allowDropInsideItem: [{ type: Input }],
        allowReordering: [{ type: Input }],
        autoScroll: [{ type: Input }],
        boundary: [{ type: Input }],
        container: [{ type: Input }],
        cursorOffset: [{ type: Input }],
        data: [{ type: Input }],
        dragDirection: [{ type: Input }],
        dragTemplate: [{ type: Input }],
        dropFeedbackMode: [{ type: Input }],
        filter: [{ type: Input }],
        group: [{ type: Input }],
        handle: [{ type: Input }],
        onAdd: [{ type: Input }],
        onDragChange: [{ type: Input }],
        onDragEnd: [{ type: Input }],
        onDragMove: [{ type: Input }],
        onDragStart: [{ type: Input }],
        onRemove: [{ type: Input }],
        onReorder: [{ type: Input }],
        scrollSensitivity: [{ type: Input }],
        scrollSpeed: [{ type: Input }],
        showDragIcons: [{ type: Input }]
    };
    return DxoRowDraggingComponent;
}(NestedOption));
export { DxoRowDraggingComponent };
var DxoRowDraggingModule = /** @class */ (function () {
    function DxoRowDraggingModule() {
    }
    DxoRowDraggingModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        DxoRowDraggingComponent
                    ],
                    exports: [
                        DxoRowDraggingComponent
                    ],
                },] },
    ];
    return DxoRowDraggingModule;
}());
export { DxoRowDraggingModule };
